### DISCUSSION TITLE
Free-Water Analysis Gradient Table Error

### DISCUSSION THREAD

**Discussion by hookeba** [dipy/dipy#2639] [Preprocessing]
## Description

Hello DIPY experts!

I am interested in exploring free-water analysis with my rat multi-shell diffusion data.  I previously performed tensor analysis using FSL's DTIFIT.  To learn more about DIPY's free-water analysis algorithm, I successfully completed the relevant tutorial "Using the Free Water Elimination Model to Remove DTI Free Water Contamination" at:

https://dipy.org/documentation/1.5.0/examples_built/reconst_fwdti/#example-reconst-fwdti

During processing of my data, I encountered the following error after assigning my gradient table variable:

>>> gtab = gradient_table(bvals, bvecs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/anaconda/anaconda3/envs/dipy/lib/python3.10/site-packages/dipy/core/gradients.py", line 599, in gradient_table
    return gradient_table_from_bvals_bvecs(bvals, bvecs, big_delta=big_delta,
  File "/home/anaconda/anaconda3/envs/dipy/lib/python3.10/site-packages/dipy/core/gradients.py", line 309, in gradient_table_from_bvals_bvecs
    raise ValueError("The vectors in bvecs should be unit (The tolerance "
ValueError: The vectors in bvecs should be unit (The tolerance can be modified as an input parameter)

However, it appears that my bval and bvec structures are appropriate. Additionally, I had no issues with FSL's DTIFIT.

```python
>>> print(bvecs.shape)
(50, 3)
>>> print(bvals.shape)
(50,)
>>> print(bvecs)
[[ 0.        0.        0.      ]
 [ 0.        0.        0.      ]
 [ 0.        0.        0.      ]
 [ 0.07606   0.065198 -0.313596]
 [-0.144375 -0.266353 -0.128805]
 [ 0.308392 -0.110048  0.034096]
 [ 0.051546 -0.249169 -0.208893]
 [ 0.242596 -0.112145 -0.192222]
 [-0.268813  0.174191  0.075995]
 [-0.207388 -0.246034  0.069541]
 [-0.153695  0.244947 -0.157345]
 [ 0.179349  0.110643 -0.252924]
 [ 0.148466 -0.155742  0.24916 ]
 [-0.126602 -0.113256 -0.281999]
 [-0.18238   0.022868  0.273118]
 [ 0.        0.        0.      ]
 [ 0.        0.        0.      ]
 [ 0.567628 -0.10879  -0.165002]
 [ 0.00284   0.60074  -0.019115]
 [ 0.530667  0.28055  -0.030769]
 [ 0.215221 -0.553984  0.089684]
 [-0.598536 -0.040457 -0.037149]
 [ 0.14356   0.412121  0.413291]
 [ 0.432234 -0.406191  0.097181]
 [ 0.229218  0.555322  0.01839 ]
 [ 0.448057  0.286509 -0.280035]
 [ 0.42628   0.408775  0.111581]
 [-0.057739 -0.557394 -0.217348]
 [ 0.480464  0.038134 -0.359113]
 [ 0.050046 -0.541078  0.256889]
 [-0.45887   0.277571 -0.271393]
 [ 0.344902 -0.48446  -0.087198]
 [-0.421355  0.115288 -0.412832]
 [ 0.079989 -0.200778  0.560849]
 [-0.427854 -0.111786 -0.407071]
 [ 0.076353 -0.414045  0.428951]
 [-0.139889  0.111042  0.573901]
 [-0.116773 -0.336164  0.484376]
 [-0.262439 -0.415065  0.346568]
 [ 0.262127 -0.035898  0.539688]
 [ 0.56321   0.118251 -0.173417]
 [-0.15934  -0.541708  0.205973]
 [ 0.370458  0.320621  0.348174]
 [-0.000618 -0.284075 -0.529682]
 [-0.548982  0.037004 -0.241891]
 [ 0.527608  0.204202  0.20296 ]
 [-0.154261  0.556756  0.165796]
 [-0.319645  0.410726  0.300654]
 [-0.245383  0.283335  0.469863]
 [-0.066589 -0.042955 -0.595804]]
>>> print(bvals)
 [  21.344647   21.344647   21.344647  289.846705  294.203906  313.074533
  322.234597  319.245168  311.067071  310.479854  316.299878  322.566682
  326.822092  307.853824  332.492411   21.344647   21.344647 1016.946478
 1016.977419 1012.419087 1012.312505 1014.394466 1041.755124 1012.994647
 1013.748579 1008.904973 1014.304798 1022.303329  974.174968 1029.250707
 1013.432138 1004.466736  966.592214 1055.180386  967.116433 1043.17996
 1056.367919 1048.222656 1035.68447  1056.098182 1017.657168 1022.892782
 1035.830631  959.693622 1026.210868 1022.618699 1019.237344 1031.507124
 1046.902172  952.284515]

```

Per Issue #2201, I also attempted the following fix without success:

np.linalg.norm(bvecs, axis=1)  

I am not certain how to modify the tolerance as an input parameter.  I have attached relevant files for inspection.  I appreciate your expert assistance with resolving this issue.

Respectfully yours

hookeba

[If reporting a bug, attach the entire traceback from Python and follow the way to reproduce below]
[If proposing an enhancement/new feature, provide links to related articles, reference examples, etc.]

## Way to reproduce
[If reporting a bug, please include the following important information:]

- [x] Operating system and version (run `python -c "import platform; print(platform.platform())"`)
 Linux-3.10.0-1160.66.1.el7.x86_64-x86_64-with-glibc2.17
- [x] Python version (run `python -c "import sys; print("Python", sys.version)"`)
3.10.5
- [x] dipy version (run `python -c "import dipy; print(dipy.__version__)"`)
1.5.0
- [x] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * import numpy; print("NumPy", numpy.__version__)
    * NumPy 1.23.1
    * import scipy; print("SciPy", scipy.__version__)
    * SciPy 1.8.1
    * import nibabel; print("Nibabel", nibabel.__version__)
    * Nibabel 4.0.1
    * import h5py; print("H5py", h5py.__version__)
    * H5py 3.7.0
    * import cvxpy; print("Cvxpy", cvxpy.__version__)
    * Cvxpy 1.2.1
    * import fury; print("fury", fury.__version__)
    * fury 0.8.0
[Rat001_DTI_Data.zip](https://github.com/dipy/dipy/files/9175048/Rat001_DTI_Data.zip)

**Comment by pjsjongsung** [comment]
Hi @hookeba

np.linalg.norm(bvec, axis=1)

provided in Issue #2201 was not a fix but a way to check if all the bvecs were a unit vector.
The 'atol' parameter is used in the gradient_table function as a tolerance threshold. (i.e. 1 - np.linalg.norm should be below atol)

You could, as mentioned by #2201 increase the atol parameter to include your bvecs, though not recommended.

I cannot guarantee the effect it would have on your analysis, but you could also try normalizing the vectors to unit vectors before feeding them.

**Comment by hookeba** [comment]
Hello @pjsjongsung 

Thank you for your suggestion.  I now understand the issue with my chosen bvecs.  I agree that the best approach is to attempt to normalize them to unit vectors instead of modifying the 'atol' parameter.  I will continue with my experimentation.

Respectfully yours

@hookeba
